test/src/unit-cppapi-vfs.cc

```diff
@@ -501,3 +502,79 @@ TEST_CASE(
     }
   }
 }
+
+TEST_CASE("C++ API: VFS recursive ls", "[cppapi][vfs][ls-recursive]") {
+  using namespace tiledb;
+  Config config;
+  tiledb_ctx_t* ctx_c;
+  tiledb_vfs_t* vfs_c;
+  auto fs_vec = test::vfs_test_get_fs_vec();
+  REQUIRE(test::vfs_test_init(fs_vec, &ctx_c, &vfs_c).ok());
+  // Sets max paths returned by recursive ls. -1 retrieves all results.
+  int64_t max_paths = GENERATE(10, 50, -1);
+
+  Context ctx(ctx_c, false);
+  VFS vfs(ctx);
+  for (const auto& fs : fs_vec) {
+    sm::URI uri(fs->temp_dir());
+    auto uri_str = uri.to_string();
+    vfs.create_dir(uri_str);
+    std::vector<std::string> expected_paths;
+    // LocalFS, MemFS returns root directories.
+    if (uri.is_file() || uri.is_memfs()) {
+      expected_paths = {
+          uri_str + "d1",
+          uri_str + "d2",
+          uri_str + "d3",
+      };
+    }
+    int top_level_dirs = static_cast<int>(expected_paths.size());
+
+    // d1 and d2 contain 10 and 100 files respectively.
+    std::vector<size_t> max_files = {10, 100, 0};
+    // Create d1, d2, d3 directories.
+    // d3 will not be returned by S3, as it is an empty prefix with no objects.
+    for (size_t i = 1; i <= 3; i++) {
+      vfs.create_dir(uri_str + "d" + std::to_string(i));
+      for (size_t j = 1; j <= max_files[i - 1]; j++) {
+        auto file = uri_str + "d" + std::to_string(i) + "/test" +
+                    std::to_string(j) + ".txt";
+        vfs.touch(file);
+        expected_paths.push_back(file);
+      }
+    }
+    // Sort and trim expected vector to match VFS::ls sorted output.
+    std::sort(expected_paths.begin() + top_level_dirs, expected_paths.end());
+    if (max_paths != -1) {
+      expected_paths.resize(max_paths);
+    }
+
+    // Ensure exception is thrown for backends without recursive ls support.
+    if (!uri.is_s3() && !uri.is_file() && !uri.is_memfs()) {
+      REQUIRE_THROWS_WITH(
+          vfs.ls_recursive(uri_str, max_paths),
+          "VFS: Recursive ls over " + uri.backend_name() +
+              " storage backend is not supported.");
+      // Test next SupportedFS.
+      continue;
+    }
+
+    auto result = vfs.ls_recursive(uri_str, max_paths);
+    auto data = result.first;
+    auto offsets = result.second;
+    for (size_t i = 1; i < offsets.size(); i++) {
+      std::string path(data, offsets[i - 1], offsets[i] - offsets[i - 1]);
+      CHECK_THAT(
+          expected_paths,
+          Catch::Matchers::VectorContains(sm::URI(path).to_string()));
+    }
+
+    // If max_paths is -1 all results should be returned.
+    // S3 won't return prefixes without objects. +3 for posix d1/, d2/, and d3/.
```

Comment:
Having an API that only sometimes returns directories seems bug prone. I'd either change it to only return files for the memfs/file URIs or do the leg work processing s3 results to insert directory entries. I'd personally lean towards files only at this point, but only narrowly since it seems easier and would have fewer subtleties on what is and isn't a directory in S3 land.